feat(backup): L2 constructs

arnulfojr opened the pull request:
closes https://github.com/aws/aws-cdk/issues/4930

- [x] PR title
- [x] unit tests
- [ ] integs
- [ ] Docs

This is my first time using the CDK so feedback is welcome!

## Testing

* Added unit tests according to the standards established by the linting tools.
## Documentation

* Added documentation for the L2 constructs, as established in the contributing guidelines, although some review would be nice to have as well.
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

<!-- 
Please read the contribution guidelines and follow the pull-request checklist:
https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md
 -->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=acb3a5de-db61-4d41-9cca-947645d1ffa2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7a777fb8-853c-4891-a457-3c7dccf728fa%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1f6746c3-3817-448b-9a3a-62011548cc3e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

arnulfojr commented:
Seems to be failing due to aws-events not being yet published with version `1.16.1`, I updated my branch to match latest commit in master

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=72702c86-a6e1-4855-b3e6-f851aa84b56e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=363f65ea-9755-4af9-8630-daeab0dbad37%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

eladb commented:
Hey, this looks like an awesome (and quality) contribution but I am afraid it will take us a while to get back to you with a review since we are a bit busy with re:Invent activities, but this is on our radar and we'll take the time to provide feedback.

Thanks again!

arnulfojr commented:
hey @eladb let me know when you have an estimated time for a review :)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=372a9c7c-f18f-4b08-a3ff-4b90d96194ca%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

arnulfojr commented on packages/@aws-cdk/aws-backup/README.md:
````diff
@@ -21,3 +21,88 @@ This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aw
+
+```ts
+new backup.BackupVault(this, 'MyVault', {
+  backupVaultName: 'MyVault',
````
The `backupVaultName` is required

arnulfojr commented on packages/@aws-cdk/aws-backup/README.md:
````diff
@@ -21,3 +21,88 @@ This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aw
+```ts
+new backup.BackupSelection(this, 'MySelections', {
+  backupPlan: backupPlan,
+  resources: dynamoTables.map(table => table.tableArn),
````
I was also wondering about this, I thought of simply making it a `type Resource` and one for `type Resources` as an array but do you have something better for passing the ARNs from different  resources?

arnulfojr commented on packages/@aws-cdk/aws-backup/lib/plan.ts:
```diff
@@ -0,0 +1,256 @@
+   * The time after a backup is scheduled before a job is canceled if it doesn't start successfully.
+   * @default backup window is set to start at 5AM UTC and lasts 8 hours.
+   */
+  readonly startWindowAfter?: Duration;
```
Yea :D I had a tough time finding a different word but this comes directly from the AWS Backup Service API, which is `startWindowMinutes` which one you think it makes more sense? https://docs.aws.amazon.com/aws-backup/latest/devguide/API_BackupRuleInput.html#Backup-Type-BackupRuleInput-StartWindowMinutes

jogold replied:
```suggestion
  readonly startWindow?: Duration;
```

arnulfojr commented on packages/@aws-cdk/aws-backup/lib/plan.ts:
```diff
@@ -0,0 +1,256 @@
+  /**
+   * The name of the backup plan
+   */
+  readonly backupPlanName: string;
```
yea, it can be optional. I'll use the `uniqueId`

jogold replied:
Using `id` here as a default is fine (+ better for the console experience and length constraint)

jogold commented:
@arnulfojr are you still working on this? If not, I will be glad to take this over.

arnulfojr commented:
@jogold hey! good to hear that someone else is also interested on this PR :)
I was thinking of continuing addressing the comments this week or next week.
From your side, is it something urgent? If so I may be able to prio if it's blocking you

jogold commented:
> From your side, is it something urgent?

Nothing urgent. If you plan to work on this I will add here my suggestions/comments, I think the API can be improved.

arnulfojr commented:
@jogold sure, comments are welcome!

jogold commented on packages/@aws-cdk/aws-backup/lib/index.ts:
```diff
@@ -1,2 +1,6 @@
 // AWS::Backup CloudFormation Resources:
 export * from './backup.generated';
+
```
The L2 `exports` are usually at the top of the file

jogold commented on packages/@aws-cdk/aws-backup/lib/vault.ts:
```diff
@@ -0,0 +1,202 @@
+/**
+ * Represents a Backup vault.
+ */
+export interface IBackupVault extends IResource {
```
I would keep only `backupVaultName` in the `IBackupVault` interface. It's the only property required to deal with a (imported) vault. This will simplify import: you'll only need a `fromBackupVaultName()`

jogold commented on packages/@aws-cdk/aws-backup/lib/vault.ts:
```diff
@@ -0,0 +1,202 @@
+
+  constructor(scope: Construct, id: string, props: BackupVaultProps) {
+    super(scope, id, {
+      physicalName: props.backupVaultName
```
You can use `this.node.uniqueId` and make `backupVaultName` optional.

jogold commented on packages/@aws-cdk/aws-backup/lib/vault.ts:
```diff
@@ -0,0 +1,202 @@
+  /**
+   * Optional access policy for the vault.
+   * @see https://docs.aws.amazon.com/aws-backup/latest/devguide/creating-a-vault-access-policy.html
+   * @default Has not access policy.
```
```suggestion
   * @default - access is not restricted.
```

jogold commented on packages/@aws-cdk/aws-backup/lib/vault.ts:
```diff
@@ -0,0 +1,202 @@
+   * Notification configuration based on SNS Topics.
+   * @default configures no notifications
+   */
+  readonly notifications?: VaultNotifications;
```
I would not nest the props here, how about `notificationTopic` and `notificationEvents` as top level props and a default of all events if `notificationTopic` is defined but `notificationsEvents` is not.

(it's not because the L1 nests its props that the L2 should also do it)

jogold commented on packages/@aws-cdk/aws-backup/lib/vault.ts:
```diff
@@ -0,0 +1,202 @@
+      physicalName: props.backupVaultName
+    });
+
+    const accessPolicy = props.accessPolicy;
```
```suggestion
    const accessPolicy = props.accessPolicy && props.accessPolicy.toJSON();
```

jogold commented on packages/@aws-cdk/aws-backup/lib/vault.ts:
```diff
@@ -0,0 +1,202 @@
+    });
+  }
+
+  private renderNotifications(notifications?: VaultNotifications): CfnBackupVault.NotificationObjectTypeProperty | undefined {
```
You need to `topic.grantPublish(new iam.ServicePrincipal('backup.amazonaws.com'))` somewhere

jogold commented on packages/@aws-cdk/aws-backup/package.json:
```diff
@@ -84,10 +84,20 @@
+  "awslint": {
+    "exclude": [
+      "awslint:docs-public-apis:@aws-cdk/aws-backup.BackupSelectionProps",
+      "awslint:docs-public-apis:@aws-cdk/aws-backup.VaultNotifications"
```
Both are documented, you can drop this

jogold commented on packages/@aws-cdk/aws-backup/lib/plan.ts:
```diff
@@ -0,0 +1,256 @@
+/**
+ * Represents a Backup Plan.
+ */
+export interface IBackupPlan extends IResource {
```
Same as for `IBackupVault`, you only need `backupPlanId` in this interface.

jogold commented on packages/@aws-cdk/aws-backup/lib/plan.ts:
```diff
@@ -0,0 +1,256 @@
+  /**
+   * The backup vault where all backup resources will be stored in.
+   */
+  readonly vault: IBackupVault;
```
Why is this required? Each rule can use its own vault.

jogold commented on packages/@aws-cdk/aws-backup/lib/plan.ts:
```diff
@@ -0,0 +1,256 @@
+ * Represents the backup plan rule configuration.
+ * If a vault is provided, this will be used instead of the one from the backup plan.
+ */
+export interface PlanRule {
```
I would make a `class` of it. This allows to offer nice static methods like `public static daily(backupVault: IBackupVault)`, `public static weekly(backupVault: IBackupVault)`, `public static monthly1Year(backupVault: IBackupVault)`...

arnulfojr replied:
Just wondering what hour of the day would be the best default to execute the backups?

jogold commented on packages/@aws-cdk/aws-backup/lib/plan.ts:
```diff
@@ -0,0 +1,256 @@
+ * A Backup Plan specifies where the backup resources will be stored and
+ * what rules will the plan will follow.
+ */
+export class BackupPlan extends BackupPlanBase {
```
Here also we can have nice static methods that mimic the console experience: `public static dailyWeeklyMonthly5YearRetention(scope: Construct, id: string, backupVault: IBackupVault)` etc.

jogold commented on packages/@aws-cdk/aws-backup/lib/plan.ts:
```diff
@@ -0,0 +1,256 @@
+  /**
+   * The name of the rule.
+   */
+  readonly ruleName: string;
```
We can have a default like `${this.node.id}Rule${this.rules.length}`

jogold commented on packages/@aws-cdk/aws-backup/lib/plan.ts:
```diff
@@ -0,0 +1,256 @@
+        completionWindowMinutes: rule.completionWindow ? rule.completionWindow.toMinutes() : undefined,
+        startWindowMinutes: rule.startWindowAfter ? rule.startWindowAfter.toMinutes() : undefined,
+        lifecycle: this.parseRuleLifecycle(rule.lifecycle),
+        scheduleExpression: rule.schedule.expressionString
```
Not 100% sure here but I think you have to validate that it's a `cron(...)` and not a `rate(...)`

arnulfojr replied:
yes, if I also recall correctly, they only support cron expressions.

jogold commented on packages/@aws-cdk/aws-backup/lib/selection.ts:
```diff
@@ -0,0 +1,154 @@
+  /**
+   * Imports a backup selection based from the given attributes.
+   */
+  public static fromBackupSelectionAttributes(scope: Construct, id: string, attrs: BackupSelectionAttributes): IBackupSelection {
```
Not sure we need to be bale to import a backup selection. There's nothing you can do with it once imported.

arnulfojr replied:
I'll try to delete it, but if I remember properly I had to add it because the lint was complaining that I should add it.

jogold commented on packages/@aws-cdk/aws-backup/lib/selection.ts:
```diff
@@ -0,0 +1,154 @@
+    const role = new iam.Role(this, 'Role', {
+      assumedBy: new iam.ServicePrincipal('backup.amazonaws.com'),
+      managedPolicies: [
+        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSBackupServiceRolePolicyForBackup')
```
Do we need the `AWSBackupServiceRolePolicyForRestores` policy as well?

arnulfojr replied:
AFAIK that policy is only if a user wants to execute the restores through backup service. Since this implementation is only executing backups I'm a bit skeptical to add the policy, this would prevent unforeseen restores.
Besides that, I don't see any reason why we shouldn't add it. Let me know if you want me to add it

arnulfojr replied:
what if we add a method `allowRestore` or similar and we add the policy. this would expose 1. the API and 2. explicit consent from the user to allow this.

jogold commented on packages/@aws-cdk/aws-backup/lib/selection.ts:
```diff
@@ -0,0 +1,154 @@
+   * A list of ARNs to backup.
+   * @default to no resources.
+   */
+  readonly resources?: string[];
```
https://docs.aws.amazon.com/aws-backup/latest/devguide/whatisbackup.html#supported-resources

jogold replied:
Would be also nice to support tags (`ListOfTags`)

MrArnoldPalmer commented:
@arnulfojr Hey! Are you still actively working on this? Or can I close it for now and you can reopen when you're ready?

jogold commented:
I can take this over. @arnulfojr let me know.

arnulfojr commented:
@jogold @MrArnoldPalmer Hey! sorry I've had little spare time for me to invest in the PR.
I have locally some changes and I'll try to make the test go green and maybe reignite this PR.
Due to the COVID-19 I've been mostly busy at home setting up the some workplace to be able to WFH :D hopefully I should have some time soon-ish.

If somebody needs this ASAP, I can push my local changes as-is (I think the most beneficial would be the updated docs) and someone can pick it up. Let me know :)

MrArnoldPalmer commented:
There is no rush to finish this but if you want to push your changes for @jogold to pick up feel free.

Otherwise reopen this whenever you're ready to keep moving. I'm gonna close it just to signal we aren't actively working on it right now now.

@jogold feel free to reopen if you pick this up as well.

jogold commented:
@arnulfojr I started to work on this. I will try to push my proposal today or tomorrow; we can the discuss/review this together.

arnulfojr commented:
@jogold :D I'm working in this PR as we speak. I'll push the changes tonight but feel free to propose it

arnulfojr commented:
Hey @eladb @MrArnoldPalmer, let's follow up in @jogold 's PR.
He added support for tags which are really valuable in this service.
